# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1380917422 14400
#      Fri Oct 04 16:10:22 2013 -0400
# Node ID 1c7d768c3847f01f336b6e62ba7bd5ccf41b402e
# Parent  9d0992c6df30a54e02bab5bb6831c8f0b958164f
don't fork if there is no controlling tty

* configure.ac: Check for ctermid.
* octave.cc (octave_fork_gui): Check for controlling tty and return
false one is not found.

diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -2077,17 +2077,17 @@ AM_PROG_CC_C_O
 
 gl_INIT
 
 ### Checks for functions and variables.
 
 dnl These checks define/undefine HAVE_FUNCNAME in config.h.
 dnl Code tests HAVE_FUNCNAME and either uses function or provides workaround.
 dnl Use multiple AC_CHECKs to avoid line continuations '\' in list
-AC_CHECK_FUNCS([canonicalize_file_name dup2])
+AC_CHECK_FUNCS([canonicalize_file_name ctermid dup2])
 AC_CHECK_FUNCS([endgrent endpwent execvp expm1 expm1f fork])
 AC_CHECK_FUNCS([getegid geteuid getgid getgrent getgrgid getgrnam])
 AC_CHECK_FUNCS([getpgrp getpid getppid getpwent getpwuid getuid])
 AC_CHECK_FUNCS([isascii kill])
 AC_CHECK_FUNCS([lgamma lgammaf lgamma_r lgammaf_r])
 AC_CHECK_FUNCS([log1p log1pf pipe])
 AC_CHECK_FUNCS([realpath resolvepath roundl])
 AC_CHECK_FUNCS([select setgrent setpwent siglongjmp strsignal])
diff --git a/libinterp/octave.cc b/libinterp/octave.cc
--- a/libinterp/octave.cc
+++ b/libinterp/octave.cc
@@ -28,16 +28,17 @@ along with Octave; see the file COPYING.
 
 #include <cassert>
 #include <cstdlib>
 #include <cstring>
 #include <ctime>
 
 #include <iostream>
 
+#include <fcntl.h>
 #include <getopt.h>
 #include <sys/types.h>
 #include <unistd.h>
 
 #include "cmd-edit.h"
 #include "f77-fcn.h"
 #include "file-ops.h"
 #include "file-stat.h"
@@ -1080,17 +1081,31 @@ octave_starting_gui (void)
 {
   start_gui = check_starting_gui ();
   return start_gui;
 }
 
 int
 octave_fork_gui (void)
 {
-  return ! no_fork_option;
+  bool have_ctty = false;
+
+#if ! (defined (__WIN32__) || defined (__APPLE__)) || defined (__CYGWIN__)
+
+#if defined (HAVE_CTERMID)
+  const char *ctty = ctermid (0);
+#else
+  const char *ctty = "/dev/tty";
+#endif
+
+  have_ctty = gnulib::open (ctty, O_RDWR, 0) > 0;
+
+#endif
+
+  return (have_ctty && ! no_fork_option);
 }
 
 DEFUN (isguirunning, args, ,
   "-*- texinfo -*-\n\
 @deftypefn {Built-in Function} {} isguirunning ()\n\
 Return true if Octave is running in GUI mode and false otherwise.\n\
 @end deftypefn")
 {
